File: src/garmin_connector/device/detector.py
import os
from pathlib import Path
from dataclasses import dataclass
from typing import List, Optional
import xml.etree.ElementTree as ET
import re

@dataclass
class GarminDeviceInfo:
    model_name: str
    unit_id: Optional[str]
    software_version: Optional[str]
    part_number: Optional[str]
    mount_point: Path
    garmin_dir: Path
    newfiles_dir: Optional[Path]
    courses_dir: Optional[Path]
    activities_dir: Optional[Path]

class GarminDeviceDetector:
    @staticmethod
    def _find_candidate_roots(custom_path: Optional[str | Path] = None) -> List[Path]:
        if custom_path:
            return [Path(custom_path)]
        
        candidates = []
        uid = os.getuid()
        
        # GVFS / MTP
        gvfs_dir = Path(f"/run/user/{uid}/gvfs")
        if gvfs_dir.exists() and gvfs_dir.is_dir():
            for child in gvfs_dir.iterdir():
                if "mtp:" in child.name.lower() or "garmin" in child.name.lower():
                    candidates.append(child)
                    if child.is_dir():
                        for subchild in child.iterdir():
                            candidates.append(subchild)
                            
        # USB Mass Storage
        usb_roots = [
            Path(f"/media/{os.environ.get('USER', 'root')}"),
            Path("/media"),
            Path(f"/run/media/{os.environ.get('USER', 'root')}"),
            Path("/mnt")
        ]
        
        for root in usb_roots:
            if root.exists() and root.is_dir():
                for child in root.iterdir():
                    if child.is_dir():
                        candidates.append(child)
                        
        return candidates

    @staticmethod
    def _parse_garmin_xml(garmin_dir: Path) -> tuple[str, Optional[str], Optional[str], Optional[str]]:
        xml_names = ["GarminDevice.xml", "GARMIN.XML", "garmindevice.xml", "garmin.xml"]
        xml_file = None
        for name in xml_names:
            candidate = garmin_dir / name
            if candidate.exists() and candidate.is_file():
                xml_file = candidate
                break
                
        if not xml_file:
            return "Garmin Generic", None, None, None
            
        try:
            xml_str = xml_file.read_text(encoding="utf-8")
            xml_str = re.sub(r'\sxmlns="[^"]+"', '', xml_str, count=1)
            root = ET.fromstring(xml_str)
            
            def get_text(path):
                el = root.find(path)
                return el.text if el is not None else None
                
            model_name = get_text("Model/Description") or "Garmin Device"
            unit_id = get_text("Id") or get_text("Unit/Id")
            software_version = get_text("SoftwareVersion") or get_text("App/Version/VersionRss")
            part_number = get_text("Model/PartNumber")
            
            return model_name, unit_id, software_version, part_number
        except Exception:
            return "Garmin Device", None, None, None

    @staticmethod
    def _find_dir_case_insensitive(parent: Path, name: str) -> Optional[Path]:
        if not parent.exists() or not parent.is_dir():
            return None
        lower_name = name.lower()
        for child in parent.iterdir():
            if child.is_dir() and child.name.lower() == lower_name:
                return child
        return None

    @staticmethod
    def detect_devices(custom_path: Optional[str | Path] = None) -> List[GarminDeviceInfo]:
        candidates = GarminDeviceDetector._find_candidate_roots(custom_path)
        devices = []
        
        for cand in candidates:
            try:
                if not cand.exists() or not cand.is_dir():
                    continue
            except PermissionError:
                continue

            garmin_dir = None
            if cand.name.lower() == "garmin":
                garmin_dir = cand
            else:
                garmin_dir = GarminDeviceDetector._find_dir_case_insensitive(cand, "garmin")
                
            if not garmin_dir:
                continue
                
            mount_point = cand.parent if cand.name.lower() == "garmin" else cand
            model, uid, sw_ver, part_num = GarminDeviceDetector._parse_garmin_xml(garmin_dir)
            
            newfiles = GarminDeviceDetector._find_dir_case_insensitive(garmin_dir, "newfiles")
            courses = GarminDeviceDetector._find_dir_case_insensitive(garmin_dir, "courses")
            activities = GarminDeviceDetector._find_dir_case_insensitive(garmin_dir, "activity")
            
            if not newfiles:
                newfiles = garmin_dir / "NEWFILES"
            if not courses:
                courses = garmin_dir / "COURSES"
                
            devices.append(GarminDeviceInfo(
                model_name=model,
                unit_id=uid,
                software_version=sw_ver,
                part_number=part_num,
                mount_point=mount_point,
                garmin_dir=garmin_dir,
                newfiles_dir=newfiles,
                courses_dir=courses,
                activities_dir=activities
            ))
            
        return devices

    @staticmethod
    def get_first_device(custom_path: Optional[str | Path] = None) -> Optional[GarminDeviceInfo]:
        devices = GarminDeviceDetector.detect_devices(custom_path)
        return devices[0] if devices else None

File: internal/device/detector.go
package device

import (
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"github.com/ganeshrvel/go-mtpx"
)

type GarminDeviceInfo struct {
	ModelName       string `json:"model_name"`
	UnitID          string `json:"unit_id,omitempty"`
	SoftwareVersion string `json:"software_version,omitempty"`
	PartNumber      string `json:"part_number,omitempty"`
	MountPoint      string `json:"mount_point"`
	GarminDir       string `json:"-"`
	NewFilesDir     string `json:"-"`
	CoursesDir      string `json:"-"`
}

func getCandidateRoots() []string {
	var roots []string
	switch runtime.GOOS {
	case "linux":
		// User mounts
		if uid := os.Getuid(); uid > 0 {
			gvfsPath := fmt.Sprintf("/run/user/%d/gvfs", uid)
			entries, _ := ioutil.ReadDir(gvfsPath)
			for _, e := range entries {
				name := strings.ToLower(e.Name())
				if strings.Contains(name, "mtp:") || strings.Contains(name, "garmin") {
					roots = append(roots, filepath.Join(gvfsPath, e.Name()))
				}
			}
		}
		user := os.Getenv("USER")
		if user != "" {
			roots = append(roots, fmt.Sprintf("/media/%s", user), fmt.Sprintf("/run/media/%s", user))
		}
		roots = append(roots, "/media", "/mnt")
	case "darwin":
		roots = append(roots, "/Volumes")
	case "windows":
		for c := 'A'; c <= 'Z'; c++ {
			roots = append(roots, fmt.Sprintf("%c:\\", c))
		}
	}
	return roots
}

func DetectDevices(customPath string) ([]GarminDeviceInfo, error) {
	var devices []GarminDeviceInfo

	checkMount := func(mount string) {
		// Does it have a GARMIN dir? Or is it GARMIN?
		garminPath := ""
		if strings.EqualFold(filepath.Base(mount), "garmin") {
			garminPath = mount
		} else {
			// Find child
			entries, err := ioutil.ReadDir(mount)
			if err != nil {
				return
			}
			for _, e := range entries {
				if e.IsDir() && strings.EqualFold(e.Name(), "garmin") {
					garminPath = filepath.Join(mount, e.Name())
					break
				}
			}
		}

		if garminPath == "" {
			return
		}

		// Read GarminDevice.xml
		info := GarminDeviceInfo{
			MountPoint: mount,
			GarminDir:  garminPath,
			ModelName:  "Garmin Generic", // Fallback
		}

		xmlPath := filepath.Join(garminPath, "GarminDevice.xml") // Try exact case first
		if _, err := os.Stat(xmlPath); os.IsNotExist(err) {
			// Case insensitive search
			entries, _ := ioutil.ReadDir(garminPath)
			for _, e := range entries {
				if !e.IsDir() && strings.EqualFold(e.Name(), "garmindevice.xml") {
					xmlPath = filepath.Join(garminPath, e.Name())
					break
				}
			}
		}

		if b, err := ioutil.ReadFile(xmlPath); err == nil {
			type Model struct {
				Description string `xml:"Description"`
			}
			type Device struct {
				Model           Model  `xml:"Model"`
				Id              string `xml:"Id"`
				SoftwareVersion string `xml:"SoftwareVersion"`
				PartNumber      string `xml:"PartNumber"`
			}
			type Doc struct {
				Device Device `xml:"Device"`
			}
			var doc Doc
			// Remove namespaces for simpler parsing
			cleanXML := strings.ReplaceAll(string(b), "xmlns", "ignore")
			if err := xml.Unmarshal([]byte(cleanXML), &doc); err == nil {
				if doc.Device.Model.Description != "" {
					info.ModelName = doc.Device.Model.Description
				}
				info.UnitID = doc.Device.Id
				info.SoftwareVersion = doc.Device.SoftwareVersion
				info.PartNumber = doc.Device.PartNumber
			}
		}

		// Find Subdirs case insensitive
		entries, _ := ioutil.ReadDir(garminPath)
		for _, e := range entries {
			if e.IsDir() {
				name := strings.ToLower(e.Name())
				if name == "newfiles" {
					info.NewFilesDir = filepath.Join(garminPath, e.Name())
				} else if name == "courses" {
					info.CoursesDir = filepath.Join(garminPath, e.Name())
				}
			}
		}
		if info.NewFilesDir == "" {
			info.NewFilesDir = filepath.Join(garminPath, "NEWFILES")
		}
		if info.CoursesDir == "" {
			info.CoursesDir = filepath.Join(garminPath, "COURSES")
		}

		devices = append(devices, info)
	}

	if customPath != "" {
		checkMount(customPath)
		return devices, nil
	}

	roots := getCandidateRoots()
	for _, root := range roots {
		if runtime.GOOS != "windows" {
			entries, err := ioutil.ReadDir(root)
			if err != nil {
				continue
			}
			for _, e := range entries {
				if e.IsDir() {
					checkMount(filepath.Join(root, e.Name()))
				}
			}
		} else {
			if _, err := os.Stat(root); err == nil {
				checkMount(root)
			}
		}
	}

	
	// Also check raw MTP devices via go-mtpx (in case gvfsd-mtp is dead)
	if customPath == "" {
		dev, err := mtpx.Initialize(mtpx.Init{DebugMode: false})
		if err == nil {
			info, err2 := mtpx.FetchDeviceInfo(dev)
			if err2 == nil {
				devices = append(devices, GarminDeviceInfo{
					ModelName: "Garmin " + info.Model,
					MountPoint: "MTP:" + info.SerialNumber,
				})
			}
			mtpx.Dispose(dev)
		}
	}
	return devices, nil
}

func GetFirstDevice(customPath string) (*GarminDeviceInfo, error) {
	devices, err := DetectDevices(customPath)
	if err != nil {
		return nil, err
	}
	if len(devices) > 0 {
		return &devices[0], nil
	}
	return nil, nil
}


